src/Array.mo

```diff
@@ -602,28 +645,91 @@ module {
   /// Space: O(1)
   public func values<T>(array : [T]) : Iter.Iter<T> = array.vals();
 
+  /// Iterator provides a single method `next()`, which returns
+  /// pairs of (index, element) in order, or `null` when out of elements to iterate over.
+  ///
+  /// ```motoko include=import
+  /// let array = [10, 11, 12];
+  ///
+  /// var sum = 0;
+  /// for ((index, element) in Array.enumerate(array)) {
+  ///   sum += element;
+  /// };
+  /// sum // => 33
+  /// ```
+  ///
+  /// Runtime: O(1)
+  ///
+  /// Space: O(1)
+  public func enumerate<T>(array : [var T]) : Iter.Iter<(Nat, T)> = object {
+    let size = array.size();
+    var index = 0;
+    public func next() : ?(Nat, T) {
+      if (index > size) {
+        return null
+      };
+      let i = index;
+      index += 1;
+      ?(i, array[i])
+    }
+  };
+
+  /// Returns true if all elements in `array` satisfy the predicate function.
+  ///
+  /// ```motoko include=import
+  /// let array = [1, 2, 3, 4];
+  /// Array.all<Nat>(array, func x = x > 0) // => true
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(1)
+  ///
+  /// *Runtime and space assumes that `predicate` runs in O(1) time and space.
   public func all<T>(array : [T], predicate : T -> Bool) : Bool {
-    todo()
+    for (element in array.vals()) {
+      if (not predicate(element)) {
+        return false
+      }
+    };
+    true
   };
 
+  /// Returns true if any element in `array` satisfies the predicate function.
+  ///
+  /// ```motoko include=import
+  /// let array = [1, 2, 3, 4];
+  /// Array.any<Nat>(array, func x = x > 3) // => true
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(1)
+  ///
+  /// *Runtime and space assumes that `predicate` runs in O(1) time and space.
   public func any<T>(array : [T], predicate : T -> Bool) : Bool {
-    todo()
+    for (element in array.vals()) {
+      if (predicate(element)) {
+        return true
+      }
+    };
+    false
   };
 
-  /// Returns a new subarray from the given array provided the start index and length of elements in the subarray
+  /// Returns a new sub-array from the given array provided the start index and length of elements in the sub-array.
   ///
   /// Limitations: Traps if the start index + length is greater than the size of the array
   ///
   /// ```motoko include=import
   ///
-  /// let array = [1,2,3,4,5];
+  /// let array = [1, 2, 3, 4, 5];
   /// let subArray = Array.subArray<Nat>(array, 2, 3);
   /// ```
   /// Runtime: O(length)
   ///
   /// Space: O(length)
   public func subArray<T>(array : [T], start : Nat, length : Nat) : [T] {
```

Comment:
I wonder if we want this to have the same parameters as range (or maybe  variant that does).
I guess it depends on what other languages do...

Reply:
Most languages use a start/end index for this function instead of the length of the array. Because this could create breaking changes if we change the parameters, we might want to rename the function if we do that. I'll try a few different possibilities and open a PR if something seems to work nicely.